### PR TITLE
Build clickhouse-keeper with Musl

### PR DESCRIPTION
**Pull request by alexey-milovidov**
### Changelog category (leave one):
- Build/Testing/Packaging Improvement

### Changelog entry (a user-readable short description of the changes that goes to CHANGELOG.md):
Now `clickhouse-keeper` for the `x86_64` architecture is statically linked with [musl](https://musl.libc.org/) and doesn't depend on any system libraries.

### CODE REVIEW

**Comment by alexey-milovidov** [comment]
Need to merge with master after #36337.

**Comment by alexey-milovidov** [comment]
Waiting for #36361.

**Comment by alesapin** [comment]
Built without failures:
```
$ ldd ./clickhouse-keeper 
        statically linked
```
But for some reason, a lot of integration tests failed during keeper start.

**Comment by alesapin** [comment]
@mergifyio update

**Comment by alesapin** [comment]
@mergify up

**Comment by alesapin** [comment]
@Mergifyio update

**Comment by alesapin** [comment]
clear to me now -- we don't need to disable the embedded keeper build at all. We need to avoid symlink creation.

**Comment by alesapin** [comment]
Works now

**Comment by alesapin** [comment]
Waiting for all tests.

File: CMakeLists.txt [hunk 1]
```diff
@@ -257,6 +257,7 @@ endif()
 
 option(USE_DEBUG_HELPERS "Enable debug helpers" ${USE_DEBUG_HELPERS})
 option(BUILD_STANDALONE_KEEPER "Build keeper as small standalone binary" OFF)
+option(CREATE_KEEPER_SYMLINK "Create symlink for clickhouse-keeper to main server binary" ON)
```
Comment (Felixoid) [inline CMakeLists.txt:260]: I'd use here another construct
```suggestion
if (NOT BUILD_STANDALONE_KEEPER)
    option(CREATE_KEEPER_SYMLINK "Create symlink for clickhouse-keeper to main server binary" ON)
else ()
    option(CREATE_KEEPER_SYMLINK "Create symlink for clickhouse-keeper to main server binary" OFF)
endif ()
```

**Comment by alesapin** [comment]
`02180_group_by_lowcardinality` -- flaky test.

File: docker/packager/packager [hunk 2]
```diff
@@ -163,9 +163,12 @@ def parse_env_variables(
         cmake_flags.append("-DCMAKE_INSTALL_PREFIX=/usr")
         cmake_flags.append("-DCMAKE_INSTALL_SYSCONFDIR=/etc")
         cmake_flags.append("-DCMAKE_INSTALL_LOCALSTATEDIR=/var")
-        cmake_flags.append("-DBUILD_STANDALONE_KEEPER=ON")
         if is_release_build(build_type, package_type, sanitizer, split_binary):
             cmake_flags.append("-DINSTALL_STRIPPED_BINARIES=ON")
+            if not is_cross_arm:
+                result.append("BUILD_MUSL_KEEPER=1")
+            else:
+                cmake_flags.append("-DBUILD_STANDALONE_KEEPER=1")
```
Comment (Felixoid) [inline docker/packager/packager:171]: A positive condition is more readable
```suggestion
            if is_cross_arm:
                cmake_flags.append("-DBUILD_STANDALONE_KEEPER=1")
            else:
                result.append("BUILD_MUSL_KEEPER=1")
```

**Comment by alesapin** [comment]
Clear win:
image

File: programs/CMakeLists.txt [hunk 3]
```diff
@@ -7,7 +7,12 @@ include(${ClickHouse_SOURCE_DIR}/cmake/strip_binary.cmake)
 # The `clickhouse` binary is a multi purpose tool that contains multiple execution modes (client, server, etc.),
 # each of them may be built and linked as a separate library.
 # If you do not know what modes you need, turn this option OFF and enable SERVER and CLIENT only.
-option (ENABLE_CLICKHOUSE_ALL "Enable all ClickHouse modes by default" ON)
+if (NOT USE_MUSL)
+    option (ENABLE_CLICKHOUSE_ALL "Enable all ClickHouse modes by default" ON)
+else ()
+  # Only clickhouse-keeper can be built with musl currently
+  option (ENABLE_CLICKHOUSE_ALL "Enable all ClickHouse modes by default" OFF)
+endif ()
```
Comment (Felixoid) [inline programs/CMakeLists.txt:15]: Pretty much the same thing about positive conditions for me :-j
```suggestion
if (USE_MUSL)
    # Only clickhouse-keeper can be built with musl currently
    option (ENABLE_CLICKHOUSE_ALL "Enable all ClickHouse modes by default" OFF)
else ()
    option (ENABLE_CLICKHOUSE_ALL "Enable all ClickHouse modes by default" ON)
endif ()
```

File: programs/CMakeLists.txt [hunk 4]
```diff
@@ -446,13 +453,17 @@ else ()
...
+        if (NOT BUILD_STANDALONE_KEEPER AND CREATE_KEEPER_SYMLINK)
             add_custom_target (clickhouse-keeper ALL COMMAND ${CMAKE_COMMAND} -E create_symlink clickhouse clickhouse-keeper DEPENDS clickhouse)
             install (FILES "${CMAKE_CURRENT_BINARY_DIR}/clickhouse-keeper" DESTINATION ${CMAKE_INSTALL_BINDIR} COMPONENT clickhouse)
         endif()
 
-        list(APPEND CLICKHOUSE_BUNDLE clickhouse-keeper)
+        # otherwise we don't build keeper
+        if (BUILD_STANDALONE_KEEPER OR CREATE_KEEPER_SYMLINK)
+            list(APPEND CLICKHOUSE_BUNDLE clickhouse-keeper)
+        endif()
```
Comment (Felixoid) [inline programs/CMakeLists.txt:466]: It looks redundant to me. We need define ENABLE_CLICKHOUSE_KEEPER as `BUILD_STANDALONE_KEEPER OR CREATE_KEEPER_SYMLINK` above

**Review by Felixoid — CHANGES_REQUESTED** [review]
There're last things to check

**Review by Felixoid — APPROVED** [review]
let's wait for builds

**Comment by alesapin** [comment]
Finally we did it.

**Pull request merged**